### ISSUE TITLE
PPO Performance improvement by reducing the number of model calls

### ISSUE THREAD

**Issue by freQuensy23-coder** [huggingface/trl#848]
I wanted to highlight a potential optimization in the PPO trainer to improve training speed. Currently, the forward pass through the model is executed three times with the same inputs during each training step:

 - When generating the response using ppo_trainer.generate() (with the LM head and torch no grad)
 - When evaluating the value head as the critic in ppo_trainer.step() (with value head and torch.no_grad())
 - When calculating the policy gradients without torch.no_grad() to update parameters like LoRA adapters - again using the (with LM head and grad)(https://github.com/huggingface/trl/blob/eda1f36c57e95399bdbfe0a74e1bb0f9af2c9508/trl/trainer/ppo_trainer.py#L760)
Since the model is called three times on the same inputs, the training speed could potentially be improved by optimizing these repeated forward passes. One approach could be to cache the embeddings or model outputs computed in the first pass, and reuse those later in the training step.

I believe optimizing these repeated computations could significantly increase overall training throughput. Please let me know if you have any other suggestions to improve this workflow. I'm happy to open a GitHub issue as well for further discussion.

Thank you for your consideration!
 
So, i think that these model calls can be optimized to increase the speed of work ~ 2 times?

## Code example
```python
for epoch, batch in tqdm(enumerate(ppo_trainer.dataloader)):
    query_tensors = batch["input_ids"]
    # Get response from the policy model
    response = ppo_trainer.generate(query_tensors[0], **generation_kwargs)
        response_tensors.append(response.squeeze()[-gen_len:])
    batch["response"] = [tokenizer.decode(r.squeeze()) for r in response_tensors]

    # Compute sentiment score # noqa
    texts = batch["response"]
    toxicity_inputs = toxicity_tokenizer(texts, padding=True, truncation=True, return_tensors="pt").to(
        ppo_trainer.accelerator.device
    )
    logits = toxicity_model(**toxicity_inputs).logits.float()
    toxicity_labels = (logits[:, 0]).tolist()

    rewards = [torch.tensor(output) for output in toxicity_labels]

    # Run PPO step
    stats = ppo_trainer.step(query_tensors, response_tensors, rewards)
    ppo_trainer.log_stats(stats, batch, rewards)
```

**Comment by lvwerra** [comment]
Indeed, between the generation and computing the logits the first time we could do some optimization but I believe we won't be benefitting from 2x improvement and the price might be a clunkier API or complicated inner caching logic. The reason is that inside `step` there is actually no redundancy: First we need the logits before the optimization as a reference. Then we start optimizing and we need up-to-date logits as well to compare against which we need to compute freshly to be in sync with the updated model. Does that make sense?

Of course if you can propose a solution to the first redundancy that gives improvements without too much complications and some benchmarks we'd be happy to consider the change.

**Comment by ejmejm** [comment]
I have made a pull request to support this optimization: https://github.com/huggingface/trl/pull/1535. Would love to get some feedback.
